okhttp/src/main/java/com/squareup/okhttp/internal/OkHttpProtocolNegotiator.java

```diff
@@ -43,126 +43,145 @@
-    if (!android) {
-      platform.configureTlsExtensions(sslSocket, hostname, protocols);
-      return;
-    }
-
-    // Enable SNI and session tickets.
-    if (hostname != null) {
-      SET_USE_SESSION_TICKETS.invokeOptionalWithoutCheckedException(sslSocket, true);
-      SET_HOSTNAME.invokeOptionalWithoutCheckedException(sslSocket, hostname);
-    }
+  public String negotiate(
+      SSLSocket sslSocket, String hostname, List<Protocol> protocols) throws IOException {
+    configureTlsExtensions(sslSocket, hostname, protocols);
+    try {
+      // Force handshake.
+      sslSocket.startHandshake();
 
-    // Enable ALPN.
-    if (SET_ALPN_PROTOCOLS.isSupported(sslSocket)) {
-      Object[] parameters = {platform.concatLengthPrefixed(protocols)};
-      SET_ALPN_PROTOCOLS.invokeWithoutCheckedException(sslSocket, parameters);
+      String negotiatedProtocol = getSelectedProtocol(sslSocket);
+      if (negotiatedProtocol == null) {
+        throw new RuntimeException("protocol negotiation failed");
+      }
+      return negotiatedProtocol;
+    } finally {
+      PLATFORM.afterHandshake(sslSocket);
     }
   }
 
-  private static void afterHandshake(SSLSocket sslSocket) {
-    if (!android) {
-      platform.afterHandshake(sslSocket);
-    }
+  /** Configure TLS extensions. */
+  protected void configureTlsExtensions(
+      SSLSocket sslSocket, String hostname, List<Protocol> protocols) {
+    PLATFORM.configureTlsExtensions(sslSocket, hostname, protocols);
   }
 
   /** Returns the negotiated protocol, or null if no protocol was negotiated. */
-  public static String getSelectedProtocol(SSLSocket socket) {
-    if (!android) {
-      return platform.getSelectedProtocol(socket);
+  public String getSelectedProtocol(SSLSocket socket) {
+    return PLATFORM.getSelectedProtocol(socket);
+  }
+
+  private static class AndroidNegotiator extends OkHttpProtocolNegotiator {
+    // setUseSessionTickets(boolean)
+    private static final OptionalMethod<Socket> SET_USE_SESSION_TICKETS =
+        new OptionalMethod<Socket>(null, "setUseSessionTickets", Boolean.TYPE);
+    // setHostname(String)
+    private static final OptionalMethod<Socket> SET_HOSTNAME =
+        new OptionalMethod<Socket>(null, "setHostname", String.class);
+    // byte[] getAlpnSelectedProtocol()
+    private static final OptionalMethod<Socket> GET_ALPN_SELECTED_PROTOCOL =
+        new OptionalMethod<Socket>(byte[].class, "getAlpnSelectedProtocol");
+    // setAlpnSelectedProtocol(byte[])
+    private static final OptionalMethod<Socket> SET_ALPN_PROTOCOLS =
+        new OptionalMethod<Socket>(null, "setAlpnProtocols", byte[].class);
+    // byte[] getNpnSelectedProtocol()
+    private static final OptionalMethod<Socket> GET_NPN_SELECTED_PROTOCOL =
+        new OptionalMethod<Socket>(byte[].class, "getNpnSelectedProtocol");
+    private static boolean android;
+
+    /**
+     * First check if a protocol has already been selected, since it's possible that the user
```

Comment:
This seems like odd JavaDoc. It seems it would work better as just a comment within the method right above getSelectedProtocol.